steps/terraform.yaml

```diff
@@ -0,0 +1,152 @@
+      commandOptions: '-backend-config=subscription_id=${{ parameters.tfInitSub }} ${{ parameters.initCommandOptions }}'
+
+  - task: TerraformInstaller@0
+    displayName: Terraform validate
+    inputs:
+      command: validate
+      workingDirectory: '$(System.DefaultWorkingDirectory)/$(buildRepoSuffix)/components/${{ parameters.stack }}'
+
+  - task: TerraformCLI@0
+    displayName: Terraform plan ${{ parameters.stack }}
+    condition: and(succeeded(), in('${{ parameters.overrideAction }}', 'plan', 'apply'))
+    inputs:
+      command: plan
+      terraformVersion: ${{ parameters.tfversion }}
+      workingDirectory: '$(System.DefaultWorkingDirectory)/$(buildRepoSuffix)/components/${{ parameters.stack }}'
+      environmentServiceName: ${{ parameters.serviceConnection }}
+      commandOptions: '-out tfplan-${{ parameters.project }}-${{ parameters.environment }}-${{ parameters.stack }}  -var env=${{ parameters.environment }} -var project=${{ parameters.project }} -var builtFrom=$(Build.Repository.Name) -var product=${{ parameters.product }} ${{ parameters.planCommandOptions }} -var-file $(System.DefaultWorkingDirectory)/$(buildRepoSuffix)/environments/${{ parameters.environment }}/${{ parameters.environment }}.tfvars -lock=false'
+
+  - task: TerraformCLI@0
+    displayName: Terraform apply ${{ parameters.stack }}
+    condition: |
+      or(
+      and(succeeded(), eq(variables['isMain'], true), eq('${{ parameters.overrideAction }}', 'apply')),
+      and(succeeded(), eq(variables['isMain'], true), eq(variables['isTriggered'], true))
+      )
+    inputs:
+      command: apply
+      terraformVersion: ${{ parameters.tfversion }}
+      workingDirectory: '$(System.DefaultWorkingDirectory)/$(buildRepoSuffix)/components/${{ parameters.stack }}'
+      environmentServiceName: ${{ parameters.serviceConnection }}
+      commandOptions: '--auto-approve tfplan-${{ parameters.project }}-${{ parameters.environment }}-${{ parameters.stack }} ${{ parameters.applyCommandOptions }}'
+
+  - ${{ if eq(parameters.overrideAction, 'destroy') }}:
+      - task: TerraformCLI@0
+        displayName: Terraform destroy ${{ parameters.stack }}
+        condition: and(succeeded(), eq(variables['isMain'], true), eq('${{ parameters.overrideAction }}', 'destroy'))
+        inputs:
+          command: destroy
+          terraformVersion: ${{ parameters.tfversion }}
+          workingDirectory: '$(System.DefaultWorkingDirectory)/$(buildRepoSuffix)/components/${{ parameters.stack }}'
+          environmentServiceName: ${{ parameters.serviceConnection }}
+          commandOptions: '-var environment=${{ parameters.environment }} -var project=${{ parameters.project }} -var builtFrom=$(Build.Repository.Name) -var product=${{ parameters.product }} ${{ parameters.destroyCommandOptions }} -var-file $(System.DefaultWorkingDirectory)/environments/${{ parameters.environment }}/${{ parameters.environment }}.tfvars'
+
+  - task: Bash@3
+    displayName: "Remove local tfstate"
```

Comment:
Is this required?
Azure-hosted agents might not require this as every time it's a new agent.
This might be required with self-hosted ones. So should we have this as a first step to ensure that we always start clean with terraform commands? 
There may be some edge cases where this doesn't run and we get the stale tfstate files at the start of the pipeline

Reply:
I reckon this is an explicit and fail-safe cleanup against breakdown of Azure disposal of agent post-run (arguably unlikely but possible).
For self-hosted agents, I can see the benefits of running this as a a pre-init step.

Reply:
i think this was needed when running a bunch of terraform plans on the same agent, probably not needed if running in separate jobs

Reply:
now removed as discussed.